### ISSUE TITLE
NoMethodError on receiving requests when running on aarch64 platform with JRuby

### ISSUE THREAD

**Issue by DashwoodIce9** [puma/puma#3465]
**Describe the bug**
Hi Puma team,

I am running a Rails server with Puma in the cloud. I am currently running it on x86_64 hardware and was trying to migrate to aarch64. Specifically to AWS Graviton EC2 instances. With no other change being made besides the underlying hardware, the Puma server now throws a `NoMethodError` upon receiving any requests. I am running Puma using JRuby v9.3.9.0 on JDK 17.

```
# jruby -v
jruby 9.3.9.0 (2.6.8) 2022-10-24 537cd1f8bc OpenJDK 64-Bit Server VM 17.0.12+8-LTS on 17.0.12+8-LTS +jit [aarch64-linux]
```

The error -

```
Error reached top of thread-pool: undefined method `>=' for nil:NilClass (NoMethodError)
```

The initial error did not come with a stacktrace so I modified the Puma source code a little bit and recompiled the gem to print that as well. This is the stacktrace -

```
Error reached top of thread-pool: undefined method `>=' for nil:NilClass (NoMethodError)
jruby9.x/lib/ruby/gems/shared/gems/puma-6.4.2-java/lib/puma/server.rb:206:in `closed_socket?': undefined method `>=' for nil:NilClass (NoMethodError)
        from jruby9.x/lib/ruby/gems/shared/gems/puma-6.4.2-java/lib/puma/request.rb:161:in `prepare_response'
        from jruby9.x/lib/ruby/gems/shared/gems/puma-6.4.2-java/lib/puma/server.rb:577:in `response_to_error'
        from jruby9.x/lib/ruby/gems/shared/gems/puma-6.4.2-java/lib/puma/server.rb:549:in `client_error'
        from jruby9.x/lib/ruby/gems/shared/gems/puma-6.4.2-java/lib/puma/server.rb:506:in `process_client'
        from jruby9.x/lib/ruby/gems/shared/gems/puma-6.4.2-java/lib/puma/server.rb:251:in `block in run'
        from jruby9.x/lib/ruby/gems/shared/gems/puma-6.4.2-java/lib/puma/thread_pool.rb:155:in `block in spawn_thread'
```

The `state` variable [here](https://github.com/puma/puma/blame/master/lib/puma/server.rb#L201-L203) is `nil` for some reason. However, I was not able to debug why. This error is only showing up when I am passing `-J--add-opens=java.base/sun.nio.ch=org.jruby.dist -J--add-opens=java.base/java.io=org.jruby.dist` as command line arguments to the JRuby process. Without these everything works just fine. This also worked fine for x86_64. It only started breaking with aarch64. I think this is probably a JRuby issue but I am opening a bug over here to understand what exactly is Puma trying to do here that's failing, so that I can reach out to the JRuby folks with that information to debug further.

**Puma config:**

```ruby
# /tmp/config.rb

max_threads_count = 100
min_threads_count = 20
# Configure the number of threads to use to answer requests.
# https://msp-greg.github.io/puma/Puma/DSL.html#threads-instance_method
threads min_threads_count, max_threads_count

# Set the environment in which the rack's app will run. The value must be a string.
# # https://msp-greg.github.io/puma/Puma/DSL.html#environment-instance_method
environment 'production'

# Define the TCP port to bind to.
# https://msp-greg.github.io/puma/Puma/DSL.html#port-instance_method
port 8080

# How long to wait for threads to stop when shutting them down. Defaults to :forever. Specifying :immediately will cause
# ::Puma to kill the threads immediately. Otherwise the value is the number of seconds to wait.
# https://msp-greg.github.io/puma/Puma/DSL.html#force_shutdown_after-instance_method
force_shutdown_after 120

# Specify the backend for the ::IO selector.
# https://msp-greg.github.io/puma/Puma/DSL.html#io_selector_backend-instance_method
io_selector_backend :java

# Control how the remote address of the connection is set. This is configurable because to calculate the true socket
# peer address a kernel syscall is required which for very fast rack handlers slows down the handling significantly.
# https://msp-greg.github.io/puma/Puma/DSL.html#set_remote_address-instance_method
set_remote_address header: 'X-FORWARDED-FOR'
```

**To Reproduce**
Please add reproduction steps here.

```ruby
# hello.ru

run lambda { |env| [200, {"Content-Type" => "text/plain"}, ["Hello World"]] }
```

```sh
# Command line to start the server

jruby9.x/bin/jruby -J--add-opens=java.base/sun.nio.ch=org.jruby.dist -J--add-opens=java.base/java.io=org.jruby.dist jruby9.x/lib/ruby/gems/shared/bin/puma -C /tmp/config.rb /tmp/hello.ru
```

```
curl -v http://localhost:8080
*   Trying 127.0.0.1:8080...
* Connected to localhost (127.0.0.1) port 8080
> GET / HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/8.3.0
> Accept: */*
>
* Empty reply from server
* Closing connection
curl: (52) Empty reply from server
```

**Expected behavior**
Get back 200 OK response with "hello world" from the server

**Desktop (please complete the following information):**
 - OS: Amazon Linux 2
 - Puma Version: 6.4.2
 - Ruby version: JRuby 9.3.9.0
 - JDK version: 17.0.12+8-LTS
 - Platform: aarch64

**Comment by dentarg** [comment]
Doesn't _really_ sound like the same issue but for good measure can you try this patch? https://github.com/puma/puma/issues/3446#issuecomment-2294162296

**Comment by MSP-Greg** [comment]
@DashwoodIce9

The code in question is using low level socket methods to determine if the client closed the socket.  This does seem to be a bug in JRuby when setup as you've described.

If you can check if PR #3466 fixes the issue, let us know.

Also, I could probably write some basic repro code (not using Puma) that shows the error/bug.  Could you check that?  I don't have access to a linux aarch64/ARM64 system...

**Comment by DashwoodIce9** [comment]
Hi @MSP-Greg, I'll give #3466 a shot tomorrow and share how it goes. Thanks for looking into this!

**Comment by dentarg** [comment]
FYI @hendrauzia

**Comment by dentarg** [comment]
FYI @headius

**Comment by DashwoodIce9** [comment]
Hi @MSP-Greg, I tested #3466 and can confirm that it does fix the issue. Thanks!

**Comment by MSP-Greg** [comment]
@headius

> So it seems we are not properly unpacking the socket data when it is closed?

Not sure, I can't test on aarch64 linux.  But the first message states:

"the Puma server now throws a NoMethodError upon receiving any requests"

and

"This error is only showing up when I am passing `-J--add-opens=java.base/sun.nio.ch=org.jruby.dist -J--add-opens=java.base/java.io=org.jruby.dist` as command line arguments to the JRuby process."

Given the above, my PR simply addressed the `undefined method `>=' for nil:NilClass (NoMethodError)` message, which is after the `unpack`...

Thanks for helping with these Puma issues.

Bit of a storm tonite.  Have a good Labor Day wekend...

**Comment by ahorek** [comment]
this issue is expected to be resolved in JRuby (9.3.10+), see https://github.com/jruby/jruby/pull/7590

**Comment by dentarg** [comment]
I read the above comment wrong earlier, I thought we were waiting for a JRuby release. Seeing it has been fixed in JRuby already, I'm going to close this. You will need to update JRuby to avoid the problem reported here.
